cassiopeia/tools/branch_length_estimator/IIDExponentialMLE.py

```diff
@@ -0,0 +1,174 @@
+
+        # # # # # Create variables of the optimization problem # # # # #
+        r_X_t_variables = dict(
+            [
+                (node_id, cp.Variable(name=f"r_X_t_{node_id}"))
+                for node_id in tree.nodes
+            ]
+        )
+
+        # # # # # Create constraints of the optimization problem # # # # #
+        a_leaf = tree.leaves[0]
+        root = tree.root
+        root_has_time_0_constraint = [r_X_t_variables[root] == 0]
+        minimum_branch_length_constraints = [
+            r_X_t_variables[child]
+            >= r_X_t_variables[parent]
+            + minimum_branch_length * r_X_t_variables[a_leaf]
+            for (parent, child) in tree.edges
+        ]
+        ultrametric_constraints = [
+            r_X_t_variables[leaf] == r_X_t_variables[a_leaf]
+            for leaf in tree.leaves
+            if leaf != a_leaf
+        ]
+        all_constraints = (
+            root_has_time_0_constraint
+            + minimum_branch_length_constraints
+            + ultrametric_constraints
+        )
+
+        # # # # # Compute the log-likelihood # # # # #
+        log_likelihood = 0
+        for (parent, child) in tree.edges:
+            edge_length = r_X_t_variables[child] - r_X_t_variables[parent]
+            num_unmutated = len(
+                tree.get_unmutated_characters_along_edge(parent, child)
+            )
+            num_mutated = len(
+                tree.get_mutated_characters_along_edge(parent, child)
+            )
+            log_likelihood += num_unmutated * (-edge_length)
+            log_likelihood += num_mutated * cp.log(
+                1 - cp.exp(-edge_length - 1e-5)  # We add eps for stability.
+            )
+
+        # # # # # Solve the problem # # # # #
+        obj = cp.Maximize(log_likelihood)
+        prob = cp.Problem(obj, all_constraints)
+        try:
+            prob.solve(solver=solver, verbose=verbose)
+        except:  # pragma: no cover
+            raise IIDExponentialMLEError("Third-party solver failed")
+
+        # # # # # Populate the tree with the estimated branch lengths # # # # #
+        times = {
+            node: float(r_X_t_variables[node].value) for node in tree.nodes
+        }
+        # Make sure that the root has time 0 (avoid epsilons)
+        times[tree.root] = 0.0
+        # We smooth out epsilons that might make a parent's time greater
+        # than its child (which can happen if minimum_branch_length=0)
+        for (parent, child) in tree.depth_first_traverse_edges():
+            times[child] = max(times[parent], times[child])
+        tree.set_times(times)
+
+        # # # # # Extract log-likelihood # # # # #
+        log_likelihood = float(log_likelihood.value)
+        if np.isnan(log_likelihood):
+            log_likelihood = -np.inf
+        self._log_likelihood = log_likelihood
+
+        # # # # # Extract mutation rate # # # # #
+        tree_depth = tree.get_depth()
+        self._mutation_rate = tree_depth
```

Comment:
I misunderstood what this parameter is supposed to represent. Why is it the maximum depth of the tree and not the probability that a site will be cut in a given time step? Are these the same thing here?

Reply:
The property `mutation_rate` is supposed to represent the _estimated mutation rate of CRISPR/Cas9 under the assumption that the tree has depth 1_ (I will add this to the property's docstring, as it's missing). This is an implementation detail, but let me clarify why it is set to the depth of the tree at this line: Note that internally, what is solved for by the convex optimization solver is not the time of each node in the tree, but rather the time _multiplied_ by the mutation rate. As a consequence, after running the third-party solver, what we have is the time of each node _multiplied_ by the mutation rate. In particular, if we now assume that the tree has depth one, it means that the estimate for the time of a leaf (i.e. the depth of the tree) is equal to the mutation rate. I will re-structure the logic of the code to make this clearer, as I can see that now it is confusing.

Reply:
Ah that makes a lot of sense! For what it's worth, I think we were on the same page as to what the `mutation_rate` variable was supposed to store but I was unsure why you were setting it the way you were. Thanks so much for the clarification here. 

Perhaps making this a bit more clear in the code would be nice for future users struggling to understand that line of code.

Reply:
Yes, I will restructure the logic.